### PR TITLE
Retry streaming exceptions (this time for sure, Rocky!)

### LINKED ISSUE #3775: Cloud spanner client does not seem to be retrying broken streams
[Documentation](http://google-cloud-python.readthedocs.io/en/latest/spanner/snapshot-usage.html) for snapshot transactions mentions:

```
If streaming a chunk fails due to a “resumable” error, Session.read() retries the StreamingRead API reqeust, passing the resume_token from the last partial result streamed.
```

This is the correct thing to do but the client actually does not seem to be doing this. I looked around a little bit in the code and I could find anyplace where it is retrying the read with the resume token. 
This is the place where the stream would throw an error if it broke:
https://github.com/GoogleCloudPlatform/google-cloud-python/blob/9d9b6c0708ae6553458ad6107e5ad8efb23762e8/spanner/google/cloud/spanner/streamed.py#L132

 I dont see any exception handling happening here and it looks like that the error would just be propagated back to the user.

### PR DESCRIPTION
**Pull request by tseaver**
Supersedes #3930.

Closes #3775.

/cc vkedia

### CODE REVIEW

**Comment by tseaver** [comment]
@vkedia, @bjwatson PTAL

**Comment by lukesneeringer** [comment]
After @vkedia says yes, I will start the process for a Spanner release.

**Comment by bjwatson** [comment]
Thanks @lukesneeringer, @tseaver, and @jonparrott!

File: spanner/tests/unit/test_snapshot.py [hunk 1]
```diff
@@ -31,6 +33,68 @@
...
+            self._make_item(0),
+            self._make_item(1, resume_token='DEADBEEF'),
+        )
+        SECOND = (  # discarded after 503
+            self._make_item(2),
+        )
+        LAST = (
+            self._make_item(3),
+        )
+        ITEMS = FIRST + SECOND + LAST
```
Comment (vkedia) [inline spanner/tests/unit/test_snapshot.py]: unused

File: spanner/tests/unit/test_snapshot.py [hunk 2]
```diff
@@ -31,6 +33,68 @@
...
+            self._make_item(2),
+            self._make_item(3),
+        )
+        raw = _MockIterator(*ITEMS)
+        restart = mock.Mock(spec=[], return_value=raw)
+        resumable = self._call_fut(restart)
+        self.assertEqual(list(resumable), list(ITEMS))
+        restart.assert_called_once_with()
+
+    def test_iteration_w_raw_raising_unavailable(self):
```
Comment (vkedia) [inline spanner/tests/unit/test_snapshot.py]: can  you also add a test case where it raises unavailable just after getting a resume token.

File: spanner/tests/unit/test_snapshot.py (hunk 1 above)
Comment (tseaver) [reply spanner/tests/unit/test_snapshot.py]: fa7bf9e

File: spanner/tests/unit/test_snapshot.py (hunk 2 above)
Comment (tseaver) [reply spanner/tests/unit/test_snapshot.py]: fa7bf9e

**Comment by lukesneeringer** [comment]
Is this approved, @vkedia?

**Pull request merged**